### PR TITLE
[Elastic Agent] Add the ability to run the Fleet Server

### LINKED ISSUE #22686: [Elastic Agent] Embed Fleet Server as part of the Elastic Agent
Fleet-Server binary should be shipped as part of the Elastic Agent: https://github.com/elastic/fleet-server

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Adds the ability for Fleet Server to be started by Elastic Agent.

This just adds the basics for Fleet Server to be ran by Elastic Agent. This does not handle the enrollment piece of starting Fleet Server to perform the enrollment for the running Elastic Agent parent. That work will be present in a follow up branch.

This provides enough for the Fleet Server to be started by the Elastic Agent and for the release manager build of Elastic Agent to include the Fleet Server.

The Fleet Server requires a custom output configuration for elasticsearch because Fleet Server requires custom permissions and connection information before Elastic Agent has even enrolled into Fleet. This can currently be set manually in the `fleet.yml` once the Elastic Agent is enrolled in the current Kibana.

```
id: d9abff40-60bf-11eb-96c5-a7508ca0826b
fleet:
  enabled: true
  access_api_key: RHRuQlJIY0JXd3N4Q3A3TmZRbW46RVpPbHp1czVRY21zVzJXQlYtZzZhZw==
  kibana:
    protocol: http
    host: localhost:5601
    hosts:
    - localhost:5601
    timeout: 5m0s
    ssl:
      verification_mode: none
      renegotiation: never
  reporting:
    threshold: 10000
    check_frequency_sec: 30
  agent:
    id: ""
  server:
    output:
      elasticsearch:
        hosts: '${ELASTICSEARCH_HOSTS:localhost:9200}'
        username: '${ELASTICSEARCH_USERNAME:elastic}'
        password: '${ELASTICSEARCH_PASSWORD:changeme}'
```

The follow up branch will handle writing the `fleet.server.output.elasticsearch` into the `fleet.yml`.

## Why is it important?

Needed so Elastic Agent can run the Fleet server.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes #22686

## Logs

```
2021-01-28T09:14:23.382-0500    INFO    operation/operator.go:245       operation 'operation-install' skipped for fleet-server.8.0.0
2021-01-28T09:14:23.772-0500    INFO    log/reporter.go:40      2021-01-28T09:14:23-05:00: type: 'STATE': sub_type: 'STARTING' message: Application: fleet-server--8.0.0[d9abff40-60bf-11eb-96c5-a7508ca0826b]: State changed to STARTING: Starting
2021-01-28T09:14:23.780-0500    INFO    operation/monitoring.go:39      operator.handleStartSidecar: monitoring is not running and disabled, no action taken
2021-01-28T09:14:24.806-0500    INFO    log/reporter.go:40      2021-01-28T09:14:24-05:00: type: 'STATE': sub_type: 'STARTING' message: Application: fleet-server--8.0.0[d9abff40-60bf-11eb-96c5-a7508ca0826b]: State changed to STARTING: Waiting on default policy with Fleet Server integration
```

### CODE REVIEW

File: x-pack/elastic-agent/spec/fleet-server.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  - filter_values:
+      selector: inputs
+      key: enabled
+      values:
+        - true
+
+  - remove_key:
+      key: output
+
+  - select_into:
```
Comment (ruflin) [inline x-pack/elastic-agent/spec/fleet-server.yml:23]: Having the output directly as part of the input would also simplify the spec file here?

File: x-pack/elastic-agent/spec/fleet-server.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        - remove_key:
+            key: access_api_key
+        - remove_key:
+            key: kibana
+        - remove_key:
+            key: reporting
+        - remove_key:
+            key: server
+
+  - map:
```
Comment (ruflin) [inline x-pack/elastic-agent/spec/fleet-server.yml:41]: In general I would expect this spec file to only require the input routing to fleet-server. Why is this needed? Is this something we just add for every input automatically and because of this need to remove it?

File: x-pack/elastic-agent/CHANGELOG.next.asciidoc [hunk 3]
```diff
@@ -64,3 +64,4 @@
 - Log level reloadable from fleet {pull}22690[22690]
 - Push log level downstream {pull}22815[22815]
 - Add metrics collection for Agent {pull}22793[22793]
+- Add support for Fleet Server {pull}
```
Comment (michalpristas) [inline x-pack/elastic-agent/CHANGELOG.next.asciidoc:67]: please fill in PR number

**Review by michalpristas — APPROVED** [review]
change looks ok

File: x-pack/elastic-agent/spec/fleet-server.yml (hunk 1 above)
Comment (blakerouse) [reply x-pack/elastic-agent/spec/fleet-server.yml:23]: It would still need to be moved because the output is a top-level key to Fleet Server.

I also don't think we want that, because we do not want to send username/password from the input. That also would have a chicken-n-egg problem of getting that information initially.

File: x-pack/elastic-agent/spec/fleet-server.yml (hunk 2 above)
Comment (blakerouse) [reply x-pack/elastic-agent/spec/fleet-server.yml:41]: Yes every input automatically gets these items from Elastic Agent, so they need to be removed.

File: x-pack/elastic-agent/spec/fleet-server.yml (hunk 1 above)
Comment (ruflin) [reply x-pack/elastic-agent/spec/fleet-server.yml:23]: This goes back to a previous discussion. I think the fleet-server config should be designed and optimised for what is needed for Elastic Agent. So either it directly supports input as the default or it maps it internally. But the translation should not be up to the Elastic Agent. It can be a temporary solution for now.

Not sure I follow the chicken-egg part around username / password? Can you elaborate?

Comment (blakerouse) [reply x-pack/elastic-agent/spec/fleet-server.yml:23]: Fleet Server does all of that. It has a top level `inputs` just like Elastic Agent. The issue is the output, which is also top level `output` and matches Elastic Agent.

The issue is on start-up there is no output, because Elastic Agent has not enrolled but Fleet Server needs to connect to an output so that the Elastic Agent can enroll.
